llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -6609,6 +6609,46 @@ void llvm::getUnderlyingObjects(const Value *V,
   } while (!Worklist.empty());
 }
 
+const Value *llvm::getUnderlyingObjectAggressive(const Value *V) {
+  const unsigned MaxVisited = 8;
+
+  SmallPtrSet<const Value *, 8> Visited;
+  SmallVector<const Value *, 8> Worklist;
+  Worklist.push_back(V);
+  const Value *Object = nullptr, *FirstObject = nullptr;
+  do {
+    const Value *P = Worklist.pop_back_val();
+    P = getUnderlyingObject(P);
+
+    if (!FirstObject)
+      FirstObject = P;
```

Comment:
Should we pull this out of the loop if possible?

This would make it clear that we shouldn't end up in a situation where e.g. getUnderlyingObject returns nullptr for the initial pointer, and then we initialize FirstObject with the underlying object for one of the arms of a select and may end up returning an underlying object that just matches one side of the select.

Reply:
Generally agree, but I'm not sure how to structure the code to do this without either repeating the getUnderlyingObject() call or duplicating a bunch of logic. Any ideas?

(Note that getUnderlyingObject cannot return null.)

Reply:
Not the nicest but:
```
 auto SetNextP = [&]() -> bool {
    if(Worklist.empty())
       return false;
    P = getUnderlyingObject(Worklist.pop_back_val());
    return true;
 };
 FirstObject = P = getUnderlyingObject(V);
 do {
     if (!Visited.insert(P).second)
     ...
  } while(SetNextP());
```

Reply:
I feel like that's harder to follow than the current logic :(

I pushed a possible variant at https://github.com/llvm/llvm-project/pull/99509/commits/f160e9ca0887ac18e60890848a9909b62e54ffa2, which does repeat the getUnderlyingObject() call, but only in the case where FirstObject is actually needed. Maybe this is fine?

Reply:
Yeah, the capturing lambda isn't nice.
I definetly do not prefer https://github.com/llvm/llvm-project/commit/f160e9ca0887ac18e60890848a9909b62e54ffa2, that really obfuscates the whole "cant be null" aspect.

Maybe just moving the flag? I.e
```
// Cant be null
FirstObject = getUnderlying(V);
// Instead of a seperate `First` bool, you could also give `P` scope outside of the loop and do `P == nullptr` as `First`.
bool First= true;
do {
  P = First ? FirstObject : getUnderlying(Worklist.pop_back_val());
  First = false;
}
```

Its basically same special case as you had but moved around to make the non-null condition clearer.

Reply:
Thanks, latest version looks good.

```
    if (!FirstObject)
       FirstObject = P;
```
can be removed from the loop?

Reply:
Oh yeah, that wasn't supposed to be there anymore... Dropped in https://github.com/llvm/llvm-project/pull/99509/commits/44714f08c499f310984ccfe16bb28d08a78ae279.